### PR TITLE
[wasm-mt] Remove two-phase suspend; fix state transitions in DS server; fix merge

### LINKED ISSUE #72857: [wasm-mt] Sampling thread requests resume for a thread with !(suspend_count > 0)
Seen here:

https://github.com/dotnet/runtime/pull/72275#issuecomment-1195165522

image

Repro (probably):

Build the runtime with `/p:WasmEnableThreads=true`; build and run the `browser-mt-eventpipe` sample.

### PR DESCRIPTION
**Pull request by lambdageek**
Grab bag of threading fixes:

1. Remove the coop two-phase transition (Partially revert 6726fae1e7fc8a8e80892f28215e18b89e3a61c7).  This was based on a misunderstanding of how Emscripten works: when the main thread is blocked in a concurrency primitive like `sem_wait`, it is still processing queued calls from other threads.  So there is no need to first suspend the worker threads and then suspend the main thread.  The implementation of two-phase suspend had a bug where it would suspend worker threads twice, making the suspend increase by 2. Since resume only decremented the count by 1, this lead to a suspend count overflow.  Fixes #72857 
2. Once the diagnostic server attaches to the runtime, switch it to GC Safe mode when it returns to JavaScript.  That is, while the diagnostic server is reacting to messages in the JS event loop, it is considered suspended by the runtime.  When it calls into C, switch to GC Unsafe (which may block if there's a STW happening).  Add thread state transitions when we come back to C, and when we wait.
3. Mark the wasm diagnostic server thread as "no sample; no gc" which means that we don't consider it for STW when there's a GC or a sample profiler active.  This is how we treat utility threads (including the non-wasm diagnostic server thread) on other platforms.
4. Fix a bad signature for `cwraps.mono_wasm_event_pipe_enable` due to a mistake in a previous merge
5. Added a new `browser-threads` sample

### CODE REVIEW

**Comment by lambdageek** [comment]
https://user-images.githubusercontent.com/480437/182699995-8e60be12-6457-4724-9606-fd11104f4128.mov

**Comment by radical** [comment]
Wasm/library tests [log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-73305-merge-133df5f86ec443ecb1/WasmTestOnBrowser-WebAssembly.Browser.HotReload.Test/1/console.82838e4c.log?helixlogtype=result):

```
ChromeDriver was started successfully.
[15:03:50] fail: [out of order message from the browser]: http://127.0.0.1:36645/dotnet.js 2:107546 "MONO_WASM: onConfigLoaded() failed: {}"
[15:03:50] fail: [out of order message from the browser]: http://127.0.0.1:36645/dotnet.js 2:107597 "MONO_WASM: Stacktrace: \n"
[15:03:50] fail: [out of order message from the browser]: http://127.0.0.1:36645/dotnet.js 2:107637 "TypeError: Cannot set properties of undefined (setting 'DOTNET_MODIFIABLE_ASSEMBLIES')\n    at Object.onConfigLoaded (http://127.0.0.1:36645/main.js:16:78)\n    at vc (http://127.0.0.1:36645/dotnet.js:3:115444)\n    at async oc (http://127.0.0.1:36645/dotnet.js:3:106316)\n    at async http://127.0.0.1:36645/dotnet.js:3:104090"
[15:03:50] fail: [out of order message from the browser]: http://127.0.0.1:36645/dotnet.js 2:15956 "Failed to load config file ./mono-config.json TypeError: Cannot set properties of undefined (setting 'DOTNET_MODIFIABLE_ASSEMBLIES')"
[15:03:50] fail: [out of order message from the browser]: http://127.0.0.1:36645/dotnet.js 2:15956 "Failed to load config file ./mono-config.json TypeError: Cannot set properties of undefined (setting 'DOTNET_MODIFIABLE_ASSEMBLIES')"
[15:03:50] fail: [out of order message from the browser]: http://127.0.0.1:36645/dotnet.js 2:17377 Uncaught (in promise)
[15:03:50] fail: [out of order message from the browser]: http://127.0.0.1:36645/dotnet.js 2:17377 Uncaught (in promise)
[15:03:50] fail: [out of order message from the browser]: http://127.0.0.1:36645/dotnet.js 2:17377 Uncaught (in promise)
[15:03:50] fail: [out of order message from the browser]: http://127.0.0.1:36645/dotnet.js 2:17377 Uncaught (in promise)
[15:03:50] fail: [out of order message from the browser]: http://127.0.0.1:36645/dotnet.js 2:17377 Uncaught (in promise)
[15:03:50] fail: [out of order message from the browser]: http://127.0.0.1:36645/dotnet.js 2:17377 Uncaught (in promise)
[15:03:50] fail: [out of order message from the browser]: http://127.0.0.1:36645/dotnet.js 2:104195 Uncaught (in promise)
[15:18:50] info: Waiting to flush log messages with a timeout of 120 secs ..
[15:18:50] fail: Flushing log messages failed with: System.OperationCanceledException: The operation was canceled.
                    at System.Threading.Channels.AsyncOperation`1.GetResult(Int16 token)
                    at System.Threading.Channels.ChannelReader`1.ReadAllAsync(CancellationToken cancellationToken)+MoveNext()
                    at System.Threading.Channels.ChannelReader`1.ReadAllAsync(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
                    at Microsoft.DotNet.XHarness.CLI.Commands.Wasm.WasmTestMessagesProcessor.RunAsync(CancellationToken token) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/WASM/WasmTestMessagesProcessor.cs:line 66
                    at Microsoft.DotNet.XHarness.CLI.Commands.Wasm.WasmTestMessagesProcessor.RunAsync(CancellationToken token) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/WASM/WasmTestMessagesProcessor.cs:line 66
                    at Microsoft.DotNet.XHarness.CLI.Commands.Wasm.WasmTestMessagesProcessor.CompleteAndFlushAsync(Nullable`1 timeout) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/WASM/WasmTestMessagesProcessor.cs:line 106. Ignoring.
[15:18:50] fail: Tests timed out. Killing driver service pid 11169
[15:18:50] fail: Application has finished with exit code TIMED_OUT but 42 was expected
```

@pavelsavara Is this the one you were fixing?
The xharness exception can be ignored, I have a fix for it locally too.

**Comment by radical** [comment]
[AOT](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-73305-merge-c40d13f96586468985/normal-System.Collections.Tests/1/console.bf595445.log?helixlogtype=result):
```
[15:35:02] info: Starting:    System.Collections.Tests.dll
[15:36:25] info: 219168800
[15:36:25] info: /datadisks/disk1/work/A2AC0902/w/A97C0939/e/wasm_build/AppBundle/dotnet.js:1881: 211542912
[15:36:25] info:       throw ptr;
[15:36:25] info:       ^
[15:36:25] info: 
[15:36:25] info: Process v8 exited with 1
```

**Comment by lambdageek** [comment]
> 

when I repro this locally, I get a stack overflow:

https://gist.github.com/lambdageek/8c07263665415982c3b793150a1626b4

(I ran the tests with `dotnet.sh build /t:Test /p:TargetOS=Browser /p:TargetArchitecture=Wasm /p:RunAOTCompilation=true`)

Checking `main`...

File: src/mono/mono/component/diagnostics_server.c [hunk 1]
```diff
@@ -122,6 +122,17 @@ ds_websocket_url;
 
 extern void mono_wasm_diagnostic_server_on_server_thread_created (char *websocket_url);
 
+/*
+ * diagnostic server pthread lifetime:
+ *
+ * server_thread called: no runtime yet
+ * server_thread calls emscripten_exit_with_live_runtime - thread is live in JS; no C stack.
+ * {runtime starts}
+ * server_thread_attach_to_runtime called from JS: MonoThradInfo* for server_thread is set,
```
Comment (kg) [inline src/mono/mono/component/diagnostics_server.c:131]:
```suggestion
 * server_thread_attach_to_runtime called from JS: MonoThreadInfo* for server_thread is set,
```

File: src/mono/mono/utils/mono-threads-state-machine.c [hunk 2]
```diff
@@ -125,7 +125,7 @@ check_thread_state (MonoThreadInfo* info)
 		g_assert (!no_safepoints);
 		/* fallthru */
 	case STATE_ASYNC_SUSPEND_REQUESTED:
-		g_assert (suspend_count > 0);
+		g_assertf (suspend_count > 0, "epxected suspend_count > 0 in current state: %s, suspend_count == %d", state_name(cur_state), suspend_count);
```
Comment (kg) [inline src/mono/mono/utils/mono-threads-state-machine.c:128]:
```suggestion
		g_assertf (suspend_count > 0, "expected suspend_count > 0 in current state: %s, suspend_count == %d", state_name(cur_state), suspend_count);
```

**Review by kg — APPROVED** [review]
I don't know the threading infra well enough to firmly approve the logic changes here, but the diffs all look clean to me and it makes sense

**Comment by lambdageek** [comment]
`Build Browser wasm Linux Release LibraryTests_AOT` `throw ptr` failure is known

**Pull request merged**